blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,149 @@
+package blockchain
+
+import (
+	"crypto/sha256"
+	"math/bits"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// parentHash returns the hash of the left and right hashes passed in.
+func parentHash(l, r chainhash.Hash) chainhash.Hash {
+	h := sha256.New()
+	h.Write(l[:])
+	h.Write(r[:])
+
+	buf := make([]byte, 0, 32)
+	first := h.Sum(buf)
+	h.Reset()
+	h.Write(first)
+	return *((*chainhash.Hash)(h.Sum(buf)))
+}
+
+// rollingMerkleTreeStore calculates the merkle root by only allocating O(logN)
+// memory where N is the total amount of leaves being included in the tree.
+type rollingMerkleTreeStore struct {
+	// roots are where the temporary merkle roots get stored while the
+	// merkle root is being calculated.
+	roots []chainhash.Hash
+
+	// numLeaves is the total leaves the store has processed.  numLeaves
+	// is required for the root calculation algorithm to work.
+	numLeaves uint64
+}
+
+// newRollingMerkleTreeStore returns a rollingMerkleTreeStore with the roots
+// allocated based on the passed in size.
+func newRollingMerkleTreeStore(size uint64) rollingMerkleTreeStore {
+	var alloc int
+	if size != 0 {
+		alloc = bits.Len64(size - 1)
+	}
+	return rollingMerkleTreeStore{roots: make([]chainhash.Hash, 0, alloc)}
+}
+
+// add adds a single hash to the merkle tree store.  Refer to algorithm 1 "AddOne" in
+// the utreexo paper (https://eprint.iacr.org/2019/611.pdf) for the exact algorithm.
+func (s *rollingMerkleTreeStore) add(add chainhash.Hash) {
+	// We can tell where the roots are by looking at the binary representation
+	// of the numLeaves.  Wherever there's a 1, there's a root.
+	//
+	// numLeaves of 8 will be '1000' in binary, so there will be one root at
+	// row 3. numLeaves of 3 will be '11' in binary, so there's two roots.  One at
+	// row 0 and one at row 1.
+	//
+	// In this loop below, we're looking for these roots by checking if there's
+	// a '1', starting from the LSB.  If there is a '1', we'll hash the root being
+	// added with that root until we hit a '0'.
+	newRoot := add
```

Comment:
> to be the root of the tallest subtree we are capable of fully computing

This doesn't sound quite right to me because it implies that there can be multiple roots that satisfy this condition. The algorithm used only results in 1 specific root that can be calculated when adding a leaf at any given leaf count.

How's this:

```
newRoot is the Hash we are adding.  If there's a sibling root we can hash with in
any given row, it'll be hashed together to form a new root a row above.
```

Reply:
Perhaps? I'm not too opinionated about what gets added here. Nothing is strictly necessary but the way I imagined it is that at any moment we have a collection of fully computable subtrees that we are holding onto that can be combined with future data. The first time we encounter a 0 is when we can no longer compute the roots of a taller subtree and so we must pause this partial computation and await more data. The other reason I put this comment in is that mutating values (which seems to be especially common in go idiom) harms readability if the name of the variable we use differs substantially from its role over its entire lifetime, this comment was also there to try and fill that gap.

Reply:
added in

> > to be the root of the tallest subtree we are capable of fully computing
> 
> This doesn't sound quite right to me because it implies that there can be multiple roots that satisfy this condition. The algorithm used only results in 1 specific root that can be calculated when adding a leaf at any given leaf count.
> 
> How's this:
> 
> ```
> newRoot is the Hash we are adding.  If there's a sibling root we can hash with in
> any given row, it'll be hashed together to form a new root a row above.
> ```

Will add in the above.

Reply:
Excellent